async/kafka/kafka.go

```diff
@@ -286,7 +236,58 @@ func (c *consumer) createInfo() {
 	c.info["start"] = c.start.String()
 }
 
-func closeConsumer(cns sarama.PartitionConsumer) {
+type handler struct {
+	consumer *consumer
+	messages chan async.Message
+}
+
+func (h handler) Setup(_ sarama.ConsumerGroupSession) error   { return nil }
+func (h handler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
+func (h handler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	ctx := context.Background()
+	for msg := range claim.Messages() {
+		log.Debugf("data received from topic %s", msg.Topic)
+		sp, chCtx := trace.ConsumerSpan(
+			ctx,
+			trace.ComponentOpName(trace.KafkaConsumerComponent, msg.Topic),
+			trace.KafkaConsumerComponent,
+			mapHeader(msg.Headers),
+		)
+
+		var ct string
+		if h.consumer.contentType != "" {
+			ct = h.consumer.contentType
+		} else {
+			ctTemp, err := determineContentType(msg.Headers)
+			if err != nil {
+				trace.SpanError(sp)
+				return errors.Wrap(err, "failed to determine content type")
+			}
+			ct = ctTemp
+		}
+
+		dec, err := async.DetermineDecoder(ct)
+		if err != nil {
+			trace.SpanError(sp)
+			return errors.Wrapf(err, "failed to determine decoder for %s", ct)
+		}
+
+		topicPartitionOffsetDiffGaugeSet(msg.Topic, msg.Partition, claim.HighWaterMarkOffset(), msg.Offset)
+
+		sess.MarkMessage(msg, "")
```

Comment:
this is tricky: if we mark the message as processed before we put it in the h.messages channel we wight acknowledge before we know that we have enqueued the message.
If we do this after we have sent it to the channel we know that it is in the channel but we do not know if it is processed yet.
the AMQP implementation adds the acknowledge mechanism to the message we create and we ack this message when it is actually processed but AMQP acknowledges every message on its own.
I think in Kafka if we acknowledge a message and the process crashes the last acknowledged message is where the consumer group begins. This means that we might lose messages...
> // Note: calling MarkOffset does not necessarily commit the offset to the backend
	// store immediately for efficiency reasons, and it may never be committed if
	// your application crashes. This means that you may end up processing the same
	// message twice, and your processing should ideally be idempotent.

Reply:
If we want to acknowledge every message on the system we have to use an unbuffered h.messages so that only one message at a time can be processed. This will be not that performant. we should be very careful with the buffer and message acknowledgment.
We should investigate if we set the buffer default 0 to support the sequential processing of the items and allow via options to increase this buffer at their own risk...